addons/bus/static/src/js/services/websocket_service.js

```diff
@@ -0,0 +1,125 @@
+import { parseCookieString } from "@web/core/browser/cookie_service";
+import { registry } from '@web/core/registry';
+
+import { WebsocketError } from '../websocket_errors';
+
+const { EventBus } = owl;
+/**
+ * Communicate with a SharedWorker in order to provide a single websocket
+ * connection shared across multiple tabs.
+ *
+ *  @emits connect
+ *  @emits disconnect
+ *  @emits reconnect
+ *  @emits reconnecting
+ *  @emits notification
+ */
+export const websocketService = {
+    dependencies: ['notification', 'user'],
+    _removeConnectionLostNotification: null,
+
+    start(env) {
+        this.env = env;
+        this._bus = new EventBus();
+        this._usesSharedWorker = 'SharedWorker' in window;
+        if ('SharedWorker' in window) {
+            this.worker = new browser.SharedWorker(
+                '/bus/static/src/js/workers/websocket_worker_script.js',
+                { name: "odoo:websocket_shared_worker" });
+            this.worker.port.start();
+            this.worker.port.addEventListener('message', this._handleMessage.bind(this));
+        } else {
+            // Fallback for browsers which do not support SharedWorker.
+            this.worker = new browser.Worker(
+                '/bus/static/src/js/workers/websocket_worker_script.js',
+                { name: "odoo:websocket_worker" });
+            this.worker.addEventListener('message', this._handleMessage.bind(this));
+        }
+        browser.addEventListener('beforeunload', this.stop.bind(this));
+        return {
+            off: this._bus.off.bind(this._bus),
+            on: this._bus.on.bind(this._bus),
+            send: this.send.bind(this),
+            stop: this.stop.bind(this),
+            trigger: this._bus.trigger.bind(this._bus),
+        };
+    },
+
+    //--------------------------------------------------------------------------
+    // PUBLIC
+    //--------------------------------------------------------------------------
+
+    /**
+     * Send a message through the websocket.
+     *
+     * @param {{path: string, kwargs: Object}} message
+     */
+    send(message) {
+        this._send('send', message);
+    },
+
+    /**
+     * Remove this tab from the websocket worker clients in order to
+     * receiving the notifications coming from the bus.
+     */
+    stop() {
+        this._send('leave');
+    },
+
+    //--------------------------------------------------------------------------
+    // PRIVATE
+    //--------------------------------------------------------------------------
+
+    /**
+    * Send a message to the worker.
+    *
+    * @param {'send' | 'leave'} action Action to be
+    * executed by the worker
+    * @param {Object|undefined} data Data required for the action to be
+    * executed
+    */
+    _send(action, data) {
+        const message = { action, data };
+        if (action === 'send') {
+            const currentCookie = document.cookie ? parseCookieString(document.cookie) : {};
```

Comment:
The cookie service only parses the cookie once. This cookie is used to link the http session with the websocket one. We have no way to find out when a user clears its cookies, the server updates its session since the websocket connection is persisted (the session is taken during the http handshake). Thus, I need to known when this cookie disappears. We could always parse document.cookie in the cookie_service but it would be an unnecessary overhead since it's only used here.